test/extensions/config_subscription/grpc/xds_failover_integration_test.cc

```diff
@@ -0,0 +1,575 @@
+      auto* ads_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      ads_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      ads_cluster->set_name("ads_cluster");
+      envoy::extensions::transport_sockets::tls::v3::UpstreamTlsContext context;
+      auto* validation_context = context.mutable_common_tls_context()->mutable_validation_context();
+      validation_context->mutable_trusted_ca()->set_filename(
+          TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcacert.pem"));
+      auto* san_matcher = validation_context->add_match_typed_subject_alt_names();
+      san_matcher->mutable_matcher()->set_suffix("lyft.com");
+      san_matcher->set_san_type(
+          envoy::extensions::transport_sockets::tls::v3::SubjectAltNameMatcher::DNS);
+      if (clientType() == Grpc::ClientType::GoogleGrpc) {
+        auto* google_grpc = grpc_service->mutable_google_grpc();
+        auto* ssl_creds = google_grpc->mutable_channel_credentials()->mutable_ssl_credentials();
+        ssl_creds->mutable_root_certs()->set_filename(
+            TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcacert.pem"));
+      }
+      ads_cluster->mutable_transport_socket()->set_name("envoy.transport_sockets.tls");
+      ads_cluster->mutable_transport_socket()->mutable_typed_config()->PackFrom(context);
+      if (failover_defined_) {
+        // Configure the API to use a failover gRPC service.
+        auto* failover_grpc_service = ads_config->add_grpc_services();
+        setGrpcService(*failover_grpc_service, "failover_ads_cluster",
+                       failover_xds_upstream_->localAddress());
+        auto* failover_ads_cluster = bootstrap.mutable_static_resources()->add_clusters();
+        failover_ads_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+        failover_ads_cluster->set_name("failover_ads_cluster");
+        envoy::extensions::transport_sockets::tls::v3::UpstreamTlsContext failover_context;
+        auto* failover_validation_context =
+            failover_context.mutable_common_tls_context()->mutable_validation_context();
+        failover_validation_context->mutable_trusted_ca()->set_filename(
+            TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcacert.pem"));
+        auto* failover_san_matcher =
+            failover_validation_context->add_match_typed_subject_alt_names();
+        failover_san_matcher->mutable_matcher()->set_suffix("lyft.com");
+        failover_san_matcher->set_san_type(
+            envoy::extensions::transport_sockets::tls::v3::SubjectAltNameMatcher::DNS);
+        if (clientType() == Grpc::ClientType::GoogleGrpc) {
+          auto* failover_google_grpc = failover_grpc_service->mutable_google_grpc();
+          auto* failover_ssl_creds =
+              failover_google_grpc->mutable_channel_credentials()->mutable_ssl_credentials();
+          failover_ssl_creds->mutable_root_certs()->set_filename(
```

Comment:
Nit: I think there is a fair bit of setup boiler plate here that could be nicer to refactor a bit.

Reply:
Ack, I agree.
I'll follow up on this in a subsequent PR.